tasks/pkg-debian.yml

```diff
@@ -31,21 +31,36 @@
   with_items:
     - "deb http://apt.datadoghq.com/ stable main"
     - "deb http://apt.datadoghq.com/ stable 6"
+    - "deb http://apt.datadoghq.com/ stable 7"
   when: not ansible_check_mode
 
-- name: Ensure Datadog repository is up-to-date
+- name: (Agent 7) Ensure Datadog repository is up-to-date
   apt_repository:
-    repo: "{{ datadog_apt_repo }}"
-    state: "{% if datadog_agent5 %}absent{% else %}present{% endif %}"
+    repo: "deb https://apt.datadoghq.com stable 7"
+    state: present
     update_cache: yes
-  when: not ansible_check_mode
+  when: (datadog_apt_repo | length == 0) and (datadog_agent_major_version == 7) and (not ansible_check_mode)
 
-- name: (agent5) Ensure Datadog repository is up-to-date
+- name: (Agent 6) Ensure Datadog repository is up-to-date
   apt_repository:
-    repo: "{{ datadog_agent5_apt_repo }}"
-    state: "{% if datadog_agent5 %}present{% else %}absent{% endif %}"
+    repo: "deb https://apt.datadoghq.com stable 6"
+    state: present
     update_cache: yes
-  when: not ansible_check_mode
+  when: (datadog_apt_repo | length == 0) and (datadog_agent_major_version == 6) and (not ansible_check_mode)
+
+- name: (Agent 5) Ensure Datadog repository is up-to-date
+  apt_repository:
+    repo: "deb https://apt.datadoghq.com stable main"
+    state: present
+    update_cache: yes
+  when: (datadog_apt_repo | length == 0) and (datadog_agent_major_version == 5) and (not ansible_check_mode)
```

Comment:
This actually would leave the previous repo (either 6 or 7) when switching major. This means that when switching from 7 to 6 without pining a version we would still install the latest of 7.

what about something more generic like this (haven't tested it):

```
￼- name: (Agent {{ item.key }} ) Ensure Datadog repository is up-to-date
￼  apt_repository:
￼    repo: "{{ item.value }}"
￼    state: "{% if item.key ==  datadog_agent_major_version and  datadog_apt_repo | lenght == 0 %}present{% else %}absent{% endif %}"
￼    update_cache: yes
￼  when: (not ansible_check_mode)
  with_dict:
    5: {{ datadog_agent5_apt_repo }}
    6: {{ datadog_agent6_apt_repo }}
    7: {{ datadog_agent7_apt_repo }}
```

Same thing for Suse and Centos.

Reply:
Good points. I think there is 2 cases here:
- the default one where they don't overwrite any repo URL: it should work outside the box
- the case where they change the URLs with custom ones. We can't do much about this case I think since ansible abstract the final files. People doing this will have to handle the revert themselves I think. WDYT ?

Reply:
That's fine, but I fail to see the purpose of having multiple `datadog_agentX_apt_repo` variables available, since you can only target one version in a role at the same time (via the `agent_major_version` variable). The only benefit I see would be if someone repeatedly switches between major agent versions while using custom repos and doesn't want to change the `datadog_apt_repo` variable each time.

Reply:
advantages are:
- it allow us to have 1 task instead of 1 per repo
- the role need to support rollback to previous major (we need to remove 7 repo when using 6 or 5 for example)

Reply:
Valid points, but isn't there a more beautiful way to achieve this than having N variables?

Reply:
some thoughts here:

* we should start using the `filename` parameter of `datadog_apt_repo` (https://docs.ansible.com/ansible/latest/modules/apt_repository_module.html) and set it to something stable (a hardcoded value like `datadog_agent` should be fine), so that the name of the `.list` file is stable whether we're installing 5, 6 or 7. This will allow rollbacks. I don't think we need to have multiple variables for `datadog_apt_repo` with this solution.
* to ease the upgrade path from the previous major version of the role, we should ensure that the `.list` file that was added by the previous major version of the role is properly absent. Otherwise once we start using `filename`, the `.list` file will have a completely different name compared to before, and users will end up with a duplicate `.list` file (which may even prevent rollbacks/introduce weird behaviors)
* I think we should have a similar logic for CentOS and SUSE (haven't looked at specifics of how it works right now though)

Reply:
About 1. 
The name of the `.list` file is already "stable", as it only depends on the url of the repo that's being added (in our case, by default the `.list` file is named `apt_datadoghq_com.list`). When setting the repo as absent, ansible searches for the list file corresponding to the repo url, and removes the line which exactly matches the repo we want absent.

Using the following implementation (one `datadog_agent_repo` var used to set a custom repo, defaults are hardcoded):
```
- name: Ensure Datadog repository is up-to-date
  apt_repository:
    repo: "{{ item.value }}"
    state: "{% if item.key == datadog_agent_major_version and datadog_apt_repo | length == 0 %}present{% else %}absent{% endif %}"
    update_cache: yes
  when: (not ansible_check_mode)
  with_dict: "{
  5: 'deb https://apt.datadoghq.com stable main',
  6: 'deb https://apt.datadoghq.com stable 6',
  7: 'deb https://apt.datadoghq.com stable 7' }"

- name: (Custom) Ensure Datadog repository is up-to-date
  apt_repository:
    repo: "{{ datadog_apt_repo }}"
    state: present
    update_cache: yes
  when: (datadog_apt_repo | length > 0) and (not ansible_check_mode)
``` 
[1]

we have the following behavior:
- if `datadog_agent_major_version == 7`: `deb https://apt.datadoghq.com stable 7` is added, `deb https://apt.datadoghq.com stable 6` and `deb https://apt.datadoghq.com stable main` are removed from the `apt_datadoghq_com.list` file.
- if `datadog_agent_major_version == 6 or 5`: same behavior, simply switch the lines added / removed
- if `datadog_apt_repo` is set: `deb https://apt.datadoghq.com stable 7`, `deb https://apt.datadoghq.com stable 6` and `deb https://apt.datadoghq.com stable main` are removed from the `apt_datadoghq_com.list` file. `{{ datadog_apt_repo }}` is added to the `{{ datadog_apt_repo_url }}.list` file.
=> The problematic scenario is thus the following: someone sets a custom repo, does a run, and then removes it; the custom repo remains here because there is no logic to remove it. However, this problem can be easily solved by using the `filename` parameter for the custom repo. Then, at the start of each run, we delete this file.

About 2.
This implementation also has the benefit of automatically working when upgrading, as we don't change the list file that is used for official repositories. The only repositories that may be left over after an upgrade are the custom repository lists, but we can't do anything about them anyway since we don't track them.

[1] Why don't we put the custom repo with the others in the same task?
That's because we need to check for emptiness before doing anything, as `apt_repository` will fail if given an empty string as `repo`.

----
>  I don't think we need to have multiple variables for datadog_apt_repo with this solution.

I think the real question here is not if we need or need not multiple `datadog_apt_repo` variables: both solutions are doable (I presented a solution which uses only one, but there is also a solution using one var per Agent version). It's: do we _want_ to have hardcoded defaults and one override variable, or one variable per Agent version?

Reply:
Olivier input looks great, having 1 file that contains our repo (even with custom URL) would simplify everything. We could then simplify things to:

- first task removing that file (`/etc/apt/sources.list.d/datadog_ansible.list` ?)
- second task adding a new repo to that file